active-rfcs/0000-tables.md

```diff
@@ -0,0 +1,514 @@
+    },
+    {
+      "id": "cell-2",
+      "x": 200,
+      "y": 100,
+      "row": 0,
+      "column": 1,
+      "width": 150,
+      "height": 50,
+      "boundElements": [{id: "text-2", type: "text"}],
+    }
+  ]},
+  {
+    "id": "row-2",
+    "index": 1,
+    "lastUpdated": 1633257617000
+    "cells:[{
+      "id": "cell-3",
+      "x": 100,
+      "y": 150,
+      "width": 120,
+      "height": 60,
+      "boundElements": [{id: "text3", type: "text"}],
+    },
+    {
+      "id": "cell-4",
+      "x": 220,
+      "y": 150,
+      "width": 130,
+      "height": 60,
+      "boundElements": [{id: "text-4", type: "text"}],
+
+    }
+  }],
+  "columns:": [{
+    "id": "column-1",
+    "index": 0,
+    "title": "Column 1",
+  },
+  {
+    "id": "column-2",
+    "index": 1,
+    "title": "Column 2",
+  }],
+  "angle": 0,
+  "strokeColor": "#000000",
+  "backgroundColor": "transparent",
+  "fillStyle": "hachure",
+  "strokeWidth": 1,
+  "strokeStyle": "solid",
+  "roughness": 0,
+  "opacity": 100,
+  "groupIds": [],
+  "seed": 1,
+  "version": 1,
+  "versionNonce": 1,
+  "isDeleted": false,
+  "boundElements": null,
+  "link": null,
+  "locked": false
+}
+```
+
+We will need to write a custom reconcilation to reconcile correctly for this case.
+
+When there are updates we will compare the `index` and if `index` is updated, it means rows and columns were reordered and compare the `lastUpdated` as well to ensure most recent update is applied to all the collaborators.
+
+Additionally each cell will also have a `timestamp` attribute if we consider virtual text elements (as mentioned in previous approach) to resolve conflicts when content updated on same cell.
+
+```js
+{
+  "id": "table-id",
+  "type": "table",
+  "x": 100,
+  "y": 100,
+  "title": "Table Title",
+  "rows": [{
+    "id": "row-1",
+    "index": 0,
+    "lastUpdated": 1633257617000
+    "cells": [{
+      "id": "cell-1",
+      "x": 100,
+      "y": 100,
+      "width": 100,
+      "height": 50,
+      "content": "Cell-11",
+      "timestamp": 1633257618000
+    },
+    {
+      "id": "cell-2",
+      "x": 200,
+      "y": 100,
+      "row": 0,
+      "column": 1,
+      "width": 150,
+      "height": 50,
+      "content": "Cell-12",
+      "timestamp": 1633257619000
+    }
+  ]},
+  {
+    "id": "row-2",
+    "index": 1,
+    "lastUpdated": 1633257620000
+    "cells:[{
+      "id": "cell-3",
+      "x": 100,
+      "y": 150,
+      "width": 120,
+      "height": 60,
+      "content": "Cell-21",
+      "timestamp": 1633257621000
+    },
+    {
+      "id": "cell-4",
+      "x": 220,
+      "y": 150,
+      "width": 130,
+      "height": 60,
+      "content": "Cell-22",
+      "timestamp": 1633257617000
+
+    }
+  }],
+  "columns:": [{
+    "id": "column-1",
+    "index": 0,
+    "title": "Column 1",
+    "timestamp": 1633257617000
+  },
+  {
+    "id": "column-2",
+    "index": 1,
+    "title": "Column 2",
+    "timestamp": 1633257617000
+  }],
```

Comment:
Moving the discussion here so it reflects all the fields (`lastUpdated, `timestamps` `index` and more) and the high-level approach.

I agree using arrays will impact the perf especially when it comes to lookup so moving to objects will be a better option hence here is the proposed structure 👇🏻 

```js
{
  "id": "table-id",
  "type": "table",
  "x": 100,
  "y": 100,
  "title": "Table Title",
  "cells": {
    "cell-11": {
      "id": "cell-11",
      "width": 100,
      "height": 50,
      "content": "Cell-11",
      "rowId": "row-1",
      "columnId": "col-1",
      "timestamp": 1633257618000
    },
    "cell-12": {
      "id": "cell-12",
      "row": 0,
      "column": 1,
      "width": 150,
      "height": 50,
      "content": "Cell-12",
       "rowId": "row-1",
      "columnId": "col-2",
      "timestamp": 1633257619000
    },
    "cell-21":{
      "id": "cell-21",
      "width": 120,
      "height": 60,
      "content": "Cell-21",
       "rowId": "row-2",
      "columnId": "col-1",
      "timestamp": 1633257621000
    },
    "cell-22": {
      "id": "cell-22",
      "width": 130,
      "height": 60,
      "content": "Cell-22",
       "rowId": "row-2",
      "columnId": "col-2",
      "timestamp": 1633257617000
    }
  },
  "rows": {
    "row-1": {
    "id": "row-1",
    "index": 0,
    "cellIds":["cell-11", "cell-12"],
    "lastUpdated": 1633257617000,
    "isDeleted": false,
    },
    "row-2": {
    "id": "row-2",
    "index": 1,
    "cellIds":["cell-21", "cell-22"],
    "lastUpdated": 1633257620000,
    "isDeleted": false,
    }
  },
  "columns:": {
    "column-1": {
    "id": "column-1",
    "index": 0,
    "title": "Column 1",
    "timestamp": 1633257617000,
     "isDeleted": false,
  },
    "column-2": {
    "id": "column-2",
    "index": 1,
    "title": "Column 2",
    "timestamp": 1633257617000,
     "isDeleted": false,
    }
  },
```

1. Each row has a mapping of `cellIds` to identify which cells are mapped to that row.
2. Depending on whether the virtual text elements approach works out, the `content` field will be replaced with `boundTextElementIds` / `boundTextId` to keep it simple and the `timestamp` field will be removed as well from each cell. Can rename the `content` to `text` to keep it simple.
3. As per `isDeleted`for tracking row and column deletion we will need it. As per `isDeleted` for each cell, I am not sure if we will need it so can revisit this later.
4. As per `width` and `height`, we need it as mentioned in RFC since each cell may have a different dimension
5. As per `x`/`y` , since `cellIds` are there for each cell so we can compute the coordinates for each cell with `cellIds` and `width` and `height`
6. I have kept `rowId` and `columnId` for each cell to identify the rows and columns for quick look up when cell updated since now it's a flatten structure. Can remove later if not needed but most likely will be needed during updates.

If the above structure looks good, I will get started on implementation and we can explore the scenarios on the way.
cc @Mrazator @dwelle

Reply:
> 1\. Each row has a mapping of cellIds to identify which cells are mapped to that row.

~~do you anticipate we'll be doing lookups from cell -> row/column to warrant the expanded data structure of `rows: {}, columns: {}` instead of simple `rows: string[][]` @Mrazator suggested?~~ 

EDIT: Actually yes, during local edits so we can easily update applicable row + column cells when source cell dimensions change. Though I assume we could do array lookup or in-memory cache and it'd be fine, too.

The way I look at it, the renderer will go row by row and look up the cell by id to render it (sow row -> cells), instead.

> 3\. As per `isDeleted` for tracking row and column deletion we will need it. As per `isDeleted` for each cell, I am not sure if we will need it so can revisit this later.

Actually scratch my previous comment, as I was considering `isDeleted` flag only from the perspective of the history. We will need the flag for collab/reconciliation. So we'll need it for each cell separately.

But why track it for whole columns/rows?

> 4\. As per width and height, we need it as mentioned in RFC since each cell may have a different dimension

True, didn't consider users can make the columns/rows bigger than the text itself. Although perhaps the MVP doesn't need to support manual resizing (but it probably won't be that hard to implement).

---

Now that I'm thinking about it more, we'll definitely need per-property syncing & reconciliation, otherwise if someone is editing two different cells in the same row/column, there would be conflicts such as when one user writes enough text it expands the row height, while the other user edits a different cell in that row.

Row/column reordering edits would have similar conflicts.

Reply:
> Actually scratch my previous comment, as I was considering isDeleted flag only from the perspective of the history. We will need the flag for collab/reconciliation. So we'll need it for each cell separately.

Why for each cell? Users can delete a row or a column but not individual cell

> But why track it for whole columns/rows?

To track the deletion of entire row / column and this will be used in undo/redo and collab reconciliation as well.

> Now that I'm thinking about it more, we'll definitely need per-property syncing & reconciliation, otherwise if someone is editing two different cells in the same row/column, there would be conflicts such as when one user writes enough text it expands the row height, while the other user edits a different cell in that row.

Per property syncing as in? With the help of `index`, `lastUpdated` in row and column, and `timestamp` for each cell we could come up with a reconciliation algorithm (theoretically as I think but will be more clear once implemented)

Reply:
> Why for each cell? Users can delete a row or a column but not individual cell
> ..
> To track the deletion of entire row / column and this will be used in undo/redo and collab reconciliation as well.

Yeah nvm, I didn't think this one through.

> Per property syncing as in? With the help of `index`, `lastUpdated` in row and column, and `timestamp` for each cell we could come up with a reconciliation algorithm (theoretically as I think but will be more clear once implemented)

So that the remote client doesn't throw away someone's `text` update just because someone (e.g. themselves) have updated `width` of that cell more recently due to table reflow.

But unless we want to keep separate `version` for each respective attribute (or ones we deem should be synced per-attribute), we'll have to come up with something else, such as in-memory versioning for attributes. But that feels brittle and would sidestep `element.version` entirely, adding more complexity.

Or we can ignore it, but then we should come up with a data structure which requires the least updates of neighboring cells during reflow.

Reply:
Yes, I get the point, this is a valid use case, I have updated the RFC with the latest structure, queries/updates so we can start the implementation. 
As per the syncing, I will come back to this later and we can decide how we want to go about it.

Reply:
Two cents from my side after flying through the discussion.

Let's keep in mind that import/export format also possibly does not have to correspond with the in-memory table structure. Trying to fit in all data which could be easily computed might be a wrong approach (i.e. `index`, `x`, `y`, `rowId`, `columnId`), as it does not allow for much flexibility later on. It could also make the element unnecessarily big (from collab/history POV) and unreadable in plain JSON. Similarly about adjusting the structure for "quick lookups" of the cells, for which we could just have synced in-memory indices - built once and updated incrementally.

We should ensure that we don't need to re-construct (and sort) the in-memory table on each modification/render. But IMHO the hardest part will be to figure out the reconciliation/history. Ideally, we would calculate the deltas by the existing history components, out of the runtime table structure. Hence the runtime structure should be array-free, for parts that we would expect to sync individually (like individual cells). It should also include `isDeleted` for rows/columns (as you have it) and avoid any relations between different elements (`i.e. boundTextElementId`) - those become hellish due to conflict resolution.

Once we have the deltas, we could enrich them with additional metadata (i.e. version, clientId) and reuse the same deltas during reconciliation. The only difference will be ensuring that (re)applying them in any order (on different clients) will converge to the same result. 

Also, we should agree on the basic expectations for the reconciliation strategy, e.g:

1. changes from different clients on independent rows/columns/cells should be reconciled (merged) without conflicts
2. clients modifying the same row/column/cell, should default to a deterministic LWW-strategy (the real question is how this should work?)
3. clients should not expect to have a (text) content reconciled within the individual cells (at least until we dive deeper into the text-based reconciliation

I strongly suggest opening a discussion about `#2` to at least come up with a conceptual idea on how we will ensure the convergency between clients - without tables being composed of multiple text elements & without the append-only log of all the deltas (as we cannot just replay them to ensure convergency).

One way is possibly keeping a `version` & `versionNonce` on the row, column & cell level, which should then be automatically contained within the deltas. However, some edge cases still come to mind:

- The issue with `versionNonce` is that a random client wins - the table could end up in an inconsistent state, as there will suddenly be data merged from multiple clients. Instead, we might want to consider having one client always win over another in case of conflicts - i.e. by having `version` & `actorId`.
- A) Client A moves row X in a different position B) Client B adds an element into row X. In the current structure both changes A) and  B) would contain change to row X - A) changing `index`, B) changing `cellIds` mapping. And reconciliation could end up with discarding A), which is incorrect. This looks like one of those cases why we should avoid (even intra-element) relations.

Reply:
Yes, most likely we will have to add per-property syncing for tables as @dwelle was also suggesting. 
Each cell property can have a `timestamp` property as well and the conflicts will be resolved based on that to ensure all collaborators' updates are applied correctly and there isn't randomness of who wins.

Let's say user A updates the width of the cell and user B updates the text of the cell, so the updates would look like 👇🏻 
```
{
  "id": "cell-11",
  {
     width: 60,
      timestamp: 1633257618000
   },
   {
       height: 50,
       timestamp: 1633257616000,
    },
   
      content: "Hello",
      timestamp: 1633257616000,
    }
}
```
And updates of user B would look like 👇🏻 

```
{
  "id": "cell-11",
  {
     width: 60,
      timestamp: 1633257616000
   },
   {
       height: 50,
       timestamp: 1633257616000,
    },
   
    content: "Hello",
    timestamp: 1633257618000,
    }
}
```
Now each property timestamp will be compared and one with a greater timestamp value will be applied for that property 
width -> User A wins
height -> Both are equal
content -> User B wins

However, I strongly think we are going towards OT so most likely we might want to try out **OT (Operational Transform)** for tables to resolve the concurrent conflicts efficiently.

Reply:
You cannot rely on regular timestamps (in such format) for synchronisation purposes (in general in a distributed system), as you cannot ensure the clocks on the client are in sync. Instead, lamports clocks (~`version`) and their combinations with `versionNonce`/`clientId` or in the form of `HLC` are usually used. Also for a traditional OT, you would need a server, which does not go well with E2EE and local-first principles.

I am also not persuaded we need syncing on the properties of the individual cells, as it will become unnecessarily complex not just for reconciliation, but also for the multiplayer history. Instead, as I suggested above, I would aim at lifting most of the metadata from the cells up. For example, if the `width` is defined in a column and the `height` is in the `row`, you don't need property-based syncing on the cell.

Surely there will be more properties on the cell level like `fontFamily`, but if we survived with element-based sync on all the other elements till now, we should be good to go for the cells as well. I would either attack property-level syncing before or after we have tables, but not as part of the tables and not just suddenly for the cells.

In its simplest form, when it comes to concurrent updates to cells, it would look something like:

User A updated `cell-11`:

```json
{
  "id": "cell-11",
  "originalText": "Hello from A",
  "fontFamily": 1,
   "version": 100
}
```

User B updated `cell-11`:
```json
{
  "id": "cell-11",
   "originalText": "Hello from B",
   "fontFamily": 2,
   "version": 100,
}
```

As we got used to now, since versions are the same, we take the one with a higher `versioNonce`. We could have `versionNonce` specified alongside the `version`. But ideally, as mentioned above. `versionNonce` should be the same for all the captured updates, so that one client consistently wins over the other, and the table data won't just become a mess.

This means we could just have one `versionNonce` on the whole `table` level, and in case of conflicts,  take all the updates from the table with higher `versionNonce`. I.e. if User A update generated a `versionNonce`: 111 and User B generate a `versionNonce`: 222, User B should win for all conflicting changes - cells with the same version. Similar versioning can be applied to `rows` and `columns`.

Reply:
Yes, a combination of `version` and `versionNonce` is what we use now as well. 

For `dimensions` we can have `width` defined on `column` and `height` defined in `row`and then the top level version and versionNonce combination should work well (apart from selecting random client when version is same and whichever version nonce is higher, but thats ok as we do that now as well)

Even if I ignore text-related properties (font size, font family etc for now where we need property-based syncing), later when we have full-fledged WYSIWYG, more properties on the cell will be added and syncing could become more challenging unless we check each type of property update or have some property based syncing algorithm - each cell has a version and version nonce (That's what feels theoretically rn).

Reply:
Rich text will be another challenge, but IMHO should not be done sooner than properties level sync. Once we have sync on properties, we will need to re-think the whole text-level reconciliation and after that, we will likely do something like https://www.inkandswitch.com/peritext/ for rich text (either ourselves or partially reuse an existing CRDT).

Reply:
Got me thinking, what's stopping us from a flat hierarchy with the additional `row` element type?

1. conceptually enforces a parent-child `table` > `row` > `cell` hierarchy
2. loading each element into an in-memory `Map` would rely on its insertions order (~ordering based on `index` just on init) and provide 0(1) lookups
3. we could keep a plain-old text element for cells (hence once we tackle rich-text / text-level-reconcilliation, tables would be solved automatically)
4. reconciliation would stay more or less untouched (for now)
5. multiplayer history would mostly be provided out of the box - at least until we recognize some conflicts worth tweaking
6. render step would skip rendering of all rows/cells (text elements) and render it with the table element (similarly as we do with bound text now) - hence keeping elements' order denormalized (similar to `frames` > `frame childs`)

![tables](https://github.com/user-attachments/assets/b8cbdb97-00a9-4d64-b00a-cc921b4d7052)